rpc, logging: add backgroundvalidation to getblockchaininfo

polespinasa opened the pull request:
`getblockchaininfo` returns `verificationprogress=1` and `initialblockdownload=false` even if there's background validation.
This PR adds information about background validation to rpc `getblockchaininfo` in a similar way to `validationprogress` does.

If assume utxo was used the output of a "sync" node performing background validation:
```
$ ./build/bin/bitcoin-cli getblockchaininfo
...
  "mediantime": 1756933740,
  "verificationprogress": 1,
  "initialblockdownload": false,
  "backgroundvalidation": {
    "snapshotheight": 880000,
    "blocks": 527589,
    "bestblockhash": "0000000000000000002326308420fa5ccd28a9155217f4d1896ab443d84148fa",
    "mediantime": 1529076654,
    "chainwork": "0000000000000000000000000000000000000000020c92fab9e5e1d8ed2d8dbc",
    "verificationprogress": 0.2815790617966284
  },
  "chainwork": "0000000000000000000000000000000000000000df97866c410b0302954919d2",
  "size_on_disk": 61198817285,

...
```

If assume utxo was not used the progress is hidden:
```
$ ./build/bin/bitcoin-cli getblockchaininfo
...
  "mediantime": 1756245700,
  "verificationprogress": 1,
  "initialblockdownload": false,
  "chainwork": "00000000000000000000000000000000000000000000000000000656d6bb052b",
  "size_on_disk": 3964972194,
...
```

The PR also updates the way we estimate the verification progress returning a 100% on the snapshot block and not on the tip as we will stop doing background validation when reaching it.

polespinasa committed:
rpc, log: add backgroundvalidation to getblockchaininfo

luke-jr reviewed (commented):
Concept ACK. Would be nice to test the active state.

lucifermmmenriquejr commented:
Thanks for asking me to reply lol 😘

EnriqueRamirez

On Wed, 27 Aug 2025, 4:50 pm Luke Dashjr, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Concept ACK. Would be nice to test the active state.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/33259#pullrequestreview-3162020387>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BSKJHMLB5H7PZXGL3ULONU33PYY4FAVCNFSM6AAAAACE4PAMZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCNRSGAZDAMZYG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

polespinasa commented:
> Would be nice to test the active state.

@luke-jr what do you mean by active state?

luke-jr commented:
When there is a background sync, not just `false`

polespinasa commented on src/rpc/blockchain.cpp:
```diff
@@ -1336,6 +1336,15 @@ RPCHelpMan getblockchaininfo()
+                {
+                    {RPCResult::Type::NUM, "blocks", /*optional=*/true, "the height of the most-work fully-background validated chain. The genesis block has height 0"},
+                    {RPCResult::Type::STR, "bestblockhash", /*optional=*/true, "the hash of the currently best block validated in background"},
+                    {RPCResult::Type::NUM_TIME, "mediantime", /*optional=*/true, "The median block time expressed in " + UNIX_EPOCH_TIME},
```
done

polespinasa commented on src/rpc/blockchain.cpp:
```diff
@@ -1376,6 +1385,18 @@ RPCHelpMan getblockchaininfo()
+        bValidation.pushKV("bestblockhash", btip.GetBlockHash().GetHex());
+        bValidation.pushKV("mediantime", btip.GetMedianTimePast());
+        bValidation.pushKV("chainwork", btip.nChainWork.GetHex());
+        bValidation.pushKV("verificationprogress", chainman.GuessVerificationProgress(&btip));
```
That's a good point, but that's actually what we are showing in the logs. Do we want to keep consistency between both? Update maybe also the logging?


```[background validation] UpdateTip: new best=00000000000000000007b796179f5abb7eed59f736e3418eb8381c29eb78527a height=744000 version=0x20400004 log2_work=93.616004 tx=747106639 date='2022-07-07T14:47:25Z' progress=0.605933 cache=358.0MiB(2606151txo)```


In `validation.cpp`
```c++
constexpr int BACKGROUND_LOG_INTERVAL = 2000;
        if (pindexNew->nHeight % BACKGROUND_LOG_INTERVAL == 0) {
            UpdateTipLog(m_chainman, coins_tip, pindexNew, __func__, "[background validation] ", "");
        }
```
```c++
static void UpdateTipLog( ... const CBlockIndex* tip,...)
{

   ...
                   FormatISO8601DateTime(tip->GetBlockTime()),
                   chainman.GuessVerificationProgress(tip),
   ...
}
```

polespinasa committed:
log: update progress calculations for background validation

updates estimations to the block snapshot instead of the main chain tip as it will stop validation after reaching that height

polespinasa commented:
a425bff75203edf855ac4d8a2fea3a0a60978147 added snapshot height

fd504bb351 modify the way verification progress is calculated for background validation -> context: https://github.com/bitcoin/bitcoin/pull/33259#discussion_r2306215222

polespinasa commented on test/functional/rpc_blockchain.py:
```diff
@@ -170,6 +171,10 @@ def _test_getblockchaininfo(self):
 
+        # check background validation
+        assert not res['backgroundvalidation']
+        assert "background" not in res.keys()
```
do we want to test that in `feature_assumeutxo.py` or in `rpc_blockchain.py`

polespinasa replied:
done in a48d92a5fb4c443dbf615715c8f25d1de9cabc9e

polespinasa commented on src/validation.h:
```diff
@@ -1163,6 +1163,9 @@ class ChainstateManager
     double GuessVerificationProgress(const CBlockIndex* pindex) const EXCLUSIVE_LOCKS_REQUIRED(GetMutex());
 
+    /** Guess background verification progress in case assume utxo was used (as a fraction between 0.0=genesis and 1.0=snapshot blocks). */
+    double GetBackgroundVerificationProgress(const CBlockIndex* pindex) const EXCLUSIVE_LOCKS_REQUIRED(GetMutex());
```
Done

danielabrozzoni reviewed (commented):
Concept ACK

I did a very rough first pass and the code looks good to me. As others have said, I would like to see a functional test testing the newly introduced fields when the background validation is in progress

polespinasa committed:
test: add background validation test for getblockchaininfo

polespinasa commented:
@danielabrozzoni @ajtowns @luke-jr 

Tests implemented, not the cleanest way for sure, but I couldn't figure out anything simpler.
Open to suggestions :)

BinaryIgor commented on test/functional/feature_assumeutxo.py:
```diff
@@ -518,6 +519,28 @@ def check_dump_output(output):
         assert_equal(utxo_info['height'], SNAPSHOT_BASE_HEIGHT)
         assert_equal(utxo_info['bestblock'], snapshot_hash)
 
+        self.log.info("Check that getblockchaininfo returns information about the background validaiton process")
```
nit: typo, should be "validation"

BinaryIgor commented on test/functional/feature_assumeutxo.py:
```diff
@@ -518,6 +519,28 @@ def check_dump_output(output):
+            "verificationprogress"
+        ]
+        res = n1.getblockchaininfo()
+        assert "background" in res.keys()
```
As this already is a rather too long test, it would be nice to move these assertions into a separate method, similarly as it done with check_tx_counts

nebula-21 reviewed (commented):
ACK  ea374564f2421df1883d16c19ac13d466adc1221

Built from source, tested, and reviewed the code. New RPC fields behave as expected.

Without loading UTXO snapshot: `backgroundvalidation=false`
<details>
<pre><code>
bitcoin-cli getblockchaininfo
  ...
  "mediantime": 1231006505,
  "verificationprogress": 7.692763291753853e-10,
  "initialblockdownload": true,
  "backgroundvalidation": false,
  "chainwork": "0000000000000000000000000000000000000000000000000000000100010001",
  "size_on_disk": 293,
  "pruned": false,
  ...
</code></pre>
</details>


With `loadtxoutset`: `backgroundvalidation=true` and background progress shown.
<details>
<pre><code>
bitcoin-cli getblockchaininfo
  ...
  "mediantime": 1737334898,
  "verificationprogress": 0.8812817484397629,
  "initialblockdownload": true,
  "backgroundvalidation": true,
  "background": {
    "snapshotheight": 880000,
    "blocks": 34031,
    "bestblockhash": "00000000b2d3088131445fb70bda85e90663c36a66d429a2089c05014b8aac2b",
    "mediantime": 1263124911,
    "chainwork": "000000000000000000000000000000000000000000000000000086355a943ee0",
    "verificationprogress": 2.990124328575242e-05
  },
  "chainwork": "0000000000000000000000000000000000000000a879619cfddf806d3bdf607b",
  "size_on_disk": 9569248,
  "pruned": false,
  ...
</code></pre>
</details>

Left a few nits in comments.

sedited commented on src/rpc/blockchain.cpp:
```diff
@@ -1391,6 +1401,21 @@ RPCHelpMan getblockchaininfo()
+    bool background_sync = historical_blocks.has_value();
+    obj.pushKV("backgroundvalidation", background_sync);
+    if (historical_blocks) {
+        UniValue bValidation(UniValue::VOBJ);
```
Can use snake case for variables as per the [dev-notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c)?

polespinasa commented on src/validation.cpp:
```diff
@@ -2889,7 +2908,8 @@ static void UpdateTipLog(
     const std::string& prefix,
-    const std::string& warning_messages) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
+    const std::string& warning_messages,
+    const bool &background_validation) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
True, removed the reference

sedited commented on src/rpc/blockchain.cpp:
```diff
@@ -1374,16 +1345,25 @@ RPCHelpMan getblockchaininfo()
+                {RPCResult::Type::OBJ, "backgroundvalidation", /*optional=*/true, "state info regarding background validation process",
+                {
+                    {RPCResult::Type::NUM, "snapshotheight", /*optional=*/true, "the height of the snapshot block"},
+                    {RPCResult::Type::NUM, "blocks", /*optional=*/true, "the height of the most-work fully-background validated chain. The genesis block has height 0"},
```
Nit: "background fully-validated chain" reads a bit clearer to me.

sedited reviewed (commented):
I think this still needs a release note.

polespinasa commented on src/rpc/blockchain.cpp:
```diff
@@ -1374,16 +1345,25 @@ RPCHelpMan getblockchaininfo()
+                    {RPCResult::Type::NUM, "blocks", /*optional=*/true, "the height of the most-work fully-background validated chain. The genesis block has height 0"},
+                    {RPCResult::Type::STR, "bestblockhash", /*optional=*/true, "the hash of the currently best block validated in background"},
+                    {RPCResult::Type::NUM_TIME, "mediantime", /*optional=*/true, "the median block time expressed in " + UNIX_EPOCH_TIME},
+                    {RPCResult::Type::NUM, "verificationprogress", /*optional=*/true, "estimate of background validation progress [0..1]"},
```
done

polespinasa committed:
release note

sedited reviewed (approved):
ACK b05466d172d6ae03e33f30c6734c920cfe1b85ca

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -541,6 +542,27 @@ def check_dump_output(output):
+        assert_equal(res["backgroundvalidation"]["mediantime"], block["mediantime"])
+        assert_equal(res["backgroundvalidation"]["chainwork"], block["chainwork"])
+        assert_greater_than(res["backgroundvalidation"]["verificationprogress"], 0)
+        assert_greater_than(1, res["backgroundvalidation"]["verificationprogress"])
```
Can you not check for a more narrow range? Also seems like a good place to use `assert_approx` instead of these two lines.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -1420,6 +1429,19 @@ RPCHelpMan getblockchaininfo()
+        UniValue background_validation(UniValue::VOBJ);
+        const CBlockIndex& btip{*CHECK_NONFATAL(historical_blocks->first)};
+        const std::optional<int> background_height{active_chainstate.SnapshotBase()->nHeight};
+        if (background_height) background_validation.pushKV("snapshotheight", *background_height);
```
Like I noted in the other comment, I think you actually want to use `TargetBlock()` here. This should explain why `snapshotheight` isn't returned in the functional test. There is really no good reason I can think of why we should have that have that number available.

fjahr reviewed (commented):
Concept ACK

fjahr reviewed (commented):
utACK 4e2c956ce8d24e74e7856d1ce35c44cc78968912

Thanks for addressing comments. Feel free to leave my remaining nits unaddressed if you don't have to push again, they are pretty minor.

polespinasa commented on src/validation.cpp:
```diff
@@ -5520,6 +5521,20 @@ double ChainstateManager::GuessVerificationProgress(const CBlockIndex* pindex) c
+double ChainstateManager::GetBackgroundVerificationProgress(const CBlockIndex& pindex) const
+{
+    AssertLockHeld(GetMutex());
+    if (!CHECK_NONFATAL(HistoricalChainstate() != nullptr)) return 0.0;
```
done

fjahr reviewed (commented):
ACK 297d24faa84080891e091906b83a99b5032d0159

sedited reviewed (approved):
ACK 297d24faa84080891e091906b83a99b5032d0159

danielabrozzoni reviewed (commented):
I left a question :) I'm almost done reviewing and testing!

danielabrozzoni reviewed (approved):
light tACK 07384c577a

Code looks good to me, I tested locally by starting up a new node, loading a utxo set, and verifying the `getblockchaininfo` output.

I'm only light acking since I'm very unfamiliar with this part of the codebase :)

sedited reviewed (approved):
ACK 25f69d970a411057681b1792c1c32084f4e49cf1

achow101 commented:
ACK 25f69d970a411057681b1792c1c32084f4e49cf1
